examples/opengl/cube/source/app.d

```diff
@@ -0,0 +1,533 @@
+module app;
+
+import vecs;
+
+import bindbc.glfw.bindstatic;
+import bindbc.glfw.types;
+import bindbc.opengl;
+import dlib.geometry;
+import dlib.math;
+
+import std.experimental.logger : errorf, trace, tracef;
+import std.string : format, fromStringz, toStringz;
+import std.typecons : Tuple, tuple;
+
+
+static immutable vertexShader = q{
+#version 330 core
+
+layout(location = 0) in vec3 pos;
+uniform mat4 view, proj, model;
+
+out vec3 pos_color;
+
+void main() {
+	gl_Position =  proj * view * model * vec4(pos.xyz, 1.0f);
+	pos_color = vec3(clamp(pos.x, 0.1, 1), clamp(pos.y, 0.1, 1), clamp(pos.z, 0.1, 1));
+}};
+
+
+static immutable fragmentShader = q{
+#version 330 core
+
+in vec3 pos_color;
+out vec4 Color;
+
+void main() {
+    Color = vec4(pos_color, 1.0f);
+}};
+
+// super simplistic input manager
+bool[int] keys;
+
+
+  // ==============
+ // # Components #
+// ==============
+
+enum Input;
+enum Graphics;
+
+struct Transform
+{
+	vec3 position;
+	vec3 size;
+	vec3 rotation;
+}
+
+struct Model
+{
+	vec3 position;
+	vec3 scale;
+	vec3 rotation;
+}
+
+struct Camera
+{
+	vec3 position;
+	Tuple!(float, "fov", float, "aspectRatio", float, "near", float, "far") frustum;
+	mat4 projection;
+	mat4 view;
+}
+
+struct Shader
+{
+	uint vao;
+	uint vbo;
+	uint ibo;
+	uint program;
+}
+
+
+  // ===========
+ // # Systems #
+// ===========
+
+void cameraSystem(World.Query!Camera query)
+{
+	import std.functional : memoize;
+	alias perpective = memoize!(perspectiveMatrix!float);
+	alias lookAt = memoize!(lookAtMatrix!float);
+
+	auto camera = query.get!Camera(query.front);
+
+	with (camera.frustum) camera.projection = perpective(fov, aspectRatio, near, far);
+	camera.view = lookAt(camera.position, camera.position + vec3(0, 0, -1), vec3(0, 1, 0));
+}
+
+void inputSystem(World.Query!Model.With!Input query)
+{
+	import std.algorithm : std_each = each;
+	import std.range : takeOne;
+
+	query.each!((_, ref model) {
+		if (keys[GLFW_KEY_W]) model.rotation.x -= 0.03;
+		if (keys[GLFW_KEY_A]) model.rotation.y += 0.03;
+		if (keys[GLFW_KEY_S]) model.rotation.x += 0.03;
+		if (keys[GLFW_KEY_D]) model.rotation.y -= 0.03;
+	});
+
+	if (keys[GLFW_KEY_ESC]) glfwGetCurrentContext.glfwSetWindowShouldClose(true);
+}
+
+void renderCubeSystem(World.Query!(Transform, Model, Graphics) queryCube, World.Query!Camera queryCamera, ref Shader shader)
+{
+	with(shader) queryCube.each!((ref transform, ref model, ref graphics)
+	{
+		  // ===================
+		 // # Cube's vertices #
+		// ===================
+
+		const halflen = vec3(transform.size.x / 2, transform.size.y / 2, transform.size.z / 2);
+		float[3 * 4 * 6] quadVBuffer;
+		with (transform) quadVBuffer = [
```

Comment:
Since you're doing very simple shading *(or even if you do Phong shading in fragment shader)*, you can reduce this to just 8 vertices and then adapt the indices. But since this is an example on ECS, I won't nitpick on the OpenGL side of things.

Reply:
I can reduce the number of vertices I use in this example you're right. I didn't know how I wanted to do the example, so I just saved future trouble and implemented the buffer with 24 vertices. But since I'm not doing any texture mapping nor messing around with colors I can just reduce the amount of data passed.

Reply:
done